draft-ietf-httpbis-http2bis.xml

```diff
@@ -2847,10 +2847,45 @@
             registered HTTP fields, see the "Hypertext Transfer Protocol (HTTP) Field Name Registry" registry maintained at <eref target="https://www.iana.org/assignments/http-fields/"/>.
           </t>
           <t>
-            Just as in HTTP/1.x, field names are strings of ASCII characters that are
-            compared in a case-insensitive fashion. Field names MUST be converted
-            to lowercase prior to their encoding in HTTP/2. A request or response containing
-            uppercase header field names MUST be treated as <xref target="malformed">malformed</xref>.
+            Field names are strings of ASCII characters that are compared in a case-insensitive
+            fashion. Field names MUST be converted to lowercase when constructing a HTTP/2
+            message. A request or response containing an uppercase character ('A' to 'Z', ASCII 0x41
+            to 0x5a) in a field name MUST be treated as <xref target="malformed">malformed</xref>.
+          </t>
+          <t>
+            HPACK is capable of carrying field names or values that are not valid in HTTP.  Though
+            HPACK can carry any octet, fields are not valid in the following cases:
+          </t>
```

Comment:
```suggestion
            Field names are strings of ASCII characters that are compared in a case-insensitive
            fashion. Field names MUST be converted to lowercase when constructing a HTTP/2
            message. A request or response containing an uppercase character ('A' to 'Z', ASCII 0x41
            to 0x5a) in a field name MUST be treated as <xref target="malformed">malformed</xref>.
          </t>
          <t>
            HPACK is capable of carrying field names or values that are not valid in HTTP, thus endpoints
            MUST perform some additional validation  and treat as <xref target="malformed">malformed</xref>
            that are not <xref target="PseudoHeaderFields">pseudo-header fields</xref> and that do not comply
            with the validation. An endpoint SHOULD, in addition to the lowercase condition above, validate
            fields against the HTTP ABNF grammar from <xref target="HTTP" section="5"/>. Any endpoint that
            does not validate against the HTTP ABNF grammar MUST at least validate against the following cases:
          </t>
```

Reply:
Leaving aside the grammar of the text for a moment (I only want to discuss that if we decide to go ahead with adding it), I think the core question is whether or not this is a SHOULD. If -semantics does not make this a SHOULD, why should we?

Reply:
Validating the HTTP fields in -semantics is a MUST, but I'm not proposing that here.   This change works equally as well with MAY instead of SHOULD. The point being making it clear that the conditions listed below are a minimum validation and that more strict validation may be done instead.

Of course it is yet clearer with ABNF:

```suggestion
            Field names are strings of ASCII characters that are compared in a case-insensitive
            fashion. Field names MUST be converted to lowercase when constructing a HTTP/2
            message. A request or response containing an uppercase character ('A' to 'Z', ASCII 0x41
            to 0x5a) in a field name MUST be treated as <xref target="malformed">malformed</xref>.
          </t>
          <t>
            Since HPACK is capable of carrying field names or values that are neither valid in HTTP
            nor secure to transmit as HTTP, fields MUST at least be minimally validated by treating as 
            <xref target="malformed">malformed</xref> any field that does not comply with the ABNF:
            <pre>
              field-name = h2-pseudo-field-name / 1*fn-char
              pseudo-field-name = ":" 1*fn-char 
              fn-char = %x21-39 / %x3b-40 / %x5B-7E ; lowercase visible symbols except ':'

              field-value = 1*fv-char
              fv-char = %x01-09 / %x0b-0c / %x0e-ff ; octets excluding null, CR and LF
            </pre>
            Alternately, fields MAY be more strictly validated and treated as 
            <xref target="malformed">malformed</xref> if they do not comply with the 
            HTTP ABNF grammar from <xref target="HTTP" section="5"/> modified for 
            lowercase field names:
            <pre>
              field-name = h2-pseudo-field-name / 1*fn-char
              pseudo-field-name = ":" 1*fn-char 
              fn-char = "!" / "#" / "$" / "%" / "&amp;" / "'" / "*" / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~" / DIGIT / %x61-7A
              
              field-value = *field-content
              field-content = field-vchar [ 1*( SP / HTAB / field-vchar ) field-vchar ]
              field-vchar = VCHAR / %x80-FF
            </pre>
          </t>
```

then delete all the bullet points below.

Reply:
> Validating the HTTP fields in -semantics is a MUST, but I'm not proposing that here.

Is it? I can’t find that text in the document, can you link me to it?

Reply:
In [section 2.4](https://www.ietf.org/archive/id/draft-ietf-httpbis-semantics-16.html#section-2.4) it says:

> A recipient MUST interpret a received protocol element according to the semantics defined for it by this specification

The ABNF is given for the fields in [section 5.](https://www.ietf.org/archive/id/draft-ietf-httpbis-semantics-16.html#name-fields), so by my reading  that would be a protocol element that MUST be interpreted according to the ABNF defined by the specification.

Reply:
I don't disagree with the reading, but section 2.4 does then caveat that MUST pretty substantially. For example, the sentence you quotes reads, in its entirety:

> A recipient MUST interpret a received protocol element according to the semantics defined for it by this specification, including extensions to this specification, unless the recipient has determined (through experience or configuration) that the sender incorrectly implements what is implied by those semantics.

On top of that unless, the next paragraph is:

> Unless noted otherwise, a recipient MAY attempt to recover a usable protocol element from an invalid construct.

I don't think this drastically changes the MUST, but I do think that we potentially want to walk the SHOULD back to a MAY here.

Reply:
@Lukasa, which "SHOULD" are you referring to?

Reply:
@martinthomson I believe @Lukasa is referring to the "SHOULD" in the first version of suggested text above:  
> An endpoint SHOULD, in addition to the lowercase condition above, validate fields against the HTTP ABNF grammar ...
If text like that is adopted, I don't particularly care if that is a SHOULD or MAY.   The key thing I suggestion is that this document explicitly calls out the possibility of validating against the full HTTP grammer in this "HttpFields" section and indicates that the minimal validation described is just that - a minimal validation.

However, I prefer the form of my second suggested text above:
> Since HPACK is capable of carrying field names or values that are neither valid in HTTP nor secure to transmit as HTTP, fields MUST at least be minimally validated ...
> Alternately, fields MAY be more strictly validated ...  with the  HTTP ABNF grammar ...

Finally, althought I prefer the ABNF usage in my second suggested text (specially as it clarifies the lowercase situation), but even if a text description is used I think that casn work with the form: "MUST minimally validate... Alternatively MAY more strictly validate...."